File: chisel/testings/median_test/src/main/scala/Median.scala
import chisel3._
// import chisel3.util._
// _root_ disambiguates from package chisel3.util.circt if user imports chisel3.util._
import _root_.circt.stage.ChiselStage

class Median extends RawModule {

    var arr = IO(Input(Vec(3, Vec(3, UInt(8.W)))))
    var med = IO(Output(UInt(8.W)))

    var f_high = Wire(Vec(3, UInt(8.W)))
    var f_med = Wire(Vec(3, UInt(8.W)))
    var f_low = Wire(Vec(3, UInt(8.W)))

    for (i <- 0 until 3) {
      comp(arr(i)(0), arr(i)(1), arr(i)(2), hi=f_high(i), lo=f_low(i), me=f_med(i))
    }

    var (s_high, s_med, s_low) = (Wire(UInt(8.W)), Wire(UInt(8.W)), Wire(UInt(8.W))) 

    comp(f_high(0), f_high(1), f_high(2), lo=s_high, mode="low")
    comp(f_med(0), f_med(1), f_med(2), me=s_med, mode="med")
    comp(f_low(0), f_low(1), f_low(2), hi=s_low, mode="high")

    comp(s_high, s_med, s_low, me=med, mode="med")

    def comp(a:UInt, b:UInt, c:UInt, hi:UInt=null, me:UInt=null, lo:UInt=null, mode:String="full") = {
        val sels = Wire(Vec(3, Bool()))
        val subs = Wire(Vec(3, UInt(8.W)))
        //Not used wires
        if ((mode == "high") || (mode == "low")) {
            sels(2) := 0.U
            subs(0) := 0.U
            subs(2) := 0.U
        }

        //first compare
        if (mode == "high") sels(0) := a < b
        else sels(0) := a > b
        if ((mode == "full") || (mode == "med")) subs(0) := Mux(sels(0), a, b)
        subs(1) := Mux(sels(0), b, a)

        //second compare
        sels(1) := subs(1) > c
        if (mode == "high") hi := Mux(sels(1), subs(1), c)
        else {
          if (mode != "low") subs(2) := Mux(sels(1), subs(1), c)
          if (mode != "med") lo := Mux(sels(1), c, subs(1))
        }

        //third compare
        if ((mode != "low") && (mode != "high")) {
          sels(2) := subs(0) > subs(2)
          if (mode != "med") hi := Mux(sels(2), subs(0), subs(2))
          me := Mux(sels(2), subs(2), subs(0))
        }
    }


}

object Main extends App {
  println(
    ChiselStage.emitSystemVerilog(
      gen = new Median,
      firtoolOpts = Array("-disable-all-randomization", "-strip-debug-info")
    )
  )
}

File: migen/median.py
from migen import *
from migen.fhdl import verilog

class median(Module):

    def __init__(self):

        self.arr = Signal(8*9, name="arr")
        self.med = Signal(8, name="med")

        rows = [[self.arr[(i*8*3):(i+1)*8*3][(j*8):(j+1)*8] for j in range(3)] for i in range(3)]

        f_high, f_med, f_low = [], [], []

        for row in rows:
            h, m, l = (Signal(8) for _ in range(3))
            self.my_comp(*row, h, l, m)
            f_high.append(h)
            f_med.append(m)
            f_low.append(l)

        s_high, s_med, s_low = [Signal(8) for _ in range(3)]

        self.my_comp(*f_high, l_out=s_high, mode='low')
        self.my_comp(*f_med, m_out=s_med, mode='med')
        self.my_comp(*f_low, h_out=s_low, mode='high')

        self.my_comp(s_high, s_med, s_low, m_out=self.med, mode='med')

        
    def my_comp(self, a, b, c, h_out=None, l_out=None, m_out=None, mode='full'):
        match mode:
            case 'full':
                sels = [Signal() for _ in range(3)]
                subs = [Signal(8) for _ in range(3)]

                self.comb += sels[0].eq(a > b)
                self.comb += If(sels[0], subs[0].eq(a)).Else(subs[0].eq(b))
                self.comb += If(sels[0], subs[1].eq(b)).Else(subs[1].eq(a))

                self.comb += sels[1].eq(subs[1] > c)
                self.comb += If(sels[1], subs[2].eq(subs[1])).Else(subs[2].eq(c))
                self.comb += If(sels[1], l_out.eq(c)).Else(l_out.eq(subs[1]))
                
                self.comb += sels[2].eq(subs[0] > subs[2])
                self.comb += If(sels[2], h_out.eq(subs[0])).Else(h_out.eq(subs[2]))
                self.comb += If(sels[2], m_out.eq(subs[2])).Else(m_out.eq(subs[0]))

            case 'low':
                sels = [Signal() for _ in range(2)]
                sub = Signal(8)

                self.comb += sels[0].eq(a > b)
                self.comb += If(sels[0], sub.eq(b)).Else(sub.eq(a))

                self.comb += sels[1].eq(sub > c)
                self.comb += If(sels[1], l_out.eq(c)).Else(l_out.eq(sub))

            case 'high':
                sels = [Signal() for _ in range(2)]
                sub = Signal(8)

                self.comb += sels[0].eq(a < b)
                self.comb += If(sels[0], sub.eq(b)).Else(sub.eq(a))

                self.comb += sels[1].eq(sub < c)
                self.comb += If(sels[1], h_out.eq(c)).Else(h_out.eq(sub))

            case 'med':
                sels = [Signal() for _ in range(3)]
                subs = [Signal(8) for _ in range(3)]

                self.comb += sels[0].eq(a > b)
                self.comb += If(sels[0], subs[0].eq(a)).Else(subs[0].eq(b))
                self.comb += If(sels[0], subs[1].eq(b)).Else(subs[1].eq(a))

                self.comb += sels[1].eq(subs[1] > c)
                self.comb += If(sels[1], subs[2].eq(subs[1])).Else(subs[2].eq(c))
                
                self.comb += sels[2].eq(subs[0] > subs[2])
                self.comb += If(sels[2], m_out.eq(subs[2])).Else(m_out.eq(subs[0]))
                

if (__name__ == "__main__"):
    
    dut = median()
    print(verilog.convert(dut, {dut.arr, dut.med}, name='median'))
